### PR TITLE
posix: fix pthread thread specific data cleanup

### PR DESCRIPTION
**Pull request by msobkowski**
The `pthread_key_data` allocated in `pthread_setspecific()` is never deallocated. Due to that `pthread_setspecific()` is returning `ENOMEM` after using it specific amount of times, depending on size of `k_malloc` `s underlying pool.

### CODE REVIEW

**Review by cfriedt — COMMENTED** [review]
Looks good - can you please also add a memory leak test (in the current test suite for pthreads keys)?

I would guess something like this should work:

* Create a pthread
* Count how many times `pthread_key_create()` and `pthread_setspecific()` can be called until getting ENOMEM.
* Terminate the thread.
* Repeat all of the above steps and ensure that the second count matches the first count.

**Comment by msobkowski** [comment]
Thanks for the review @cfriedt, we've added the test you suggested, could you take another look?

File: tests/posix/common/src/key.c [hunk 1]
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
...
+	}
+
+	return NULL;
+}
+
+ZTEST(key, test_thread_specific_data_deallocation)
+{
+	pthread_t thread;
+	int alloc_count_t0 = 0;
+	int alloc_count_t1 = 0;
```
Comment (cfriedt) [inline tests/posix/common/src/key.c:169]: With hardware stack protection, and these counters in stack memory, the thread that allocates keys will likely trigger a fault on write.

I would suggest making the variables static or moving them to the global scope (or both).

```suggestion
	static int alloc_count_t0 = 0;
	static int alloc_count_t1 = 0;
```

File: tests/posix/common/src/key.c [hunk 2]
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
...
+
+	zassert_ok(pthread_create(&thread, NULL, setspecific_thread, &alloc_count_t0),
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
+	zassert_ok(pthread_create(&thread, NULL, setspecific_thread, &alloc_count_t1),
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
+	zassert_ok(alloc_count_t0 != alloc_count_t1,
```
Comment (cfriedt) [inline tests/posix/common/src/key.c:179]:
```suggestion
	zassert_equal(alloc_count_t0, alloc_count_t1,
```

**Review by cfriedt — COMMENTED** [review]
A couple of small changes

File: tests/posix/common/src/key.c [hunk 3]
```diff
@@ -144,6 +144,42 @@ ZTEST(key, test_correct_key_is_deleted)
...
+ZTEST(key, test_thread_specific_data_deallocation)
+{
+	pthread_t thread;
+	int alloc_count_t0 = 0;
+	int alloc_count_t1 = 0;
+
+	zassert_ok(pthread_create(&thread, NULL, setspecific_thread, &alloc_count_t0),
+		"attempt to create thread failed");
+	zassert_ok(pthread_join(thread, NULL), "failed to join thread");
+
```
Comment (cfriedt) [inline tests/posix/common/src/key.c:174]: Maybe before spawning the second thread, add a `printk()` that displays how many times a new key was allocated and assigned a value.

**Review by cfriedt — APPROVED** [review]

**Comment by msobkowski** [comment]
@ycsin could you also take a look?

File: tests/posix/common/testcase.yaml [hunk 4]
```diff
@@ -67,6 +67,8 @@ tests:
       - CONFIG_DYNAMIC_THREAD=y
       - CONFIG_THREAD_STACK_INFO=y
       - CONFIG_HEAP_MEM_POOL_SIZE=16384
+      - CONFIG_POSIX_THREAD_KEYS_MAX=2048
+      - CONFIG_TEST_EXTRA_STACK_SIZE=16384
```
Comment (cfriedt) [inline tests/posix/common/testcase.yaml:71]: That's a lot of extra stack size - is that necessary?

**Review by ycsin — APPROVED** [review]

**Pull request merged**